src/reverse_proxy.rs

```diff
@@ -0,0 +1,99 @@
+use tokio::net::{TcpListener, TcpStream};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use std::net::{SocketAddr, Shutdown};
+use crate::errors::*;
+
+// using 9443 instead of 443 for https to avoid running with sudo during test
+pub const DEFAULT_PORT : u16 = 9443; 
+
+pub fn run(
+    local_addr: SocketAddr, 
+    server_ips: Vec<SocketAddr>, 
+    compress: bool, 
+    encrypt: bool,
+) -> Result<()>
+{
+    let rt = tokio::runtime::Runtime::new()
+        .chain_err(|| "failed to create tokio runtime")?;
+
+    rt.block_on(async {
+        let mut server_carousel = server_ips.iter().cycle();
+
+        println!("opening listener socket on {}", local_addr);
+
+        let listen_socket = TcpListener::bind(local_addr).await
+            .chain_err(|| 
+                format!("error opening listener socket on {}", local_addr)
+            )?;
+
+        loop {
+            let (mut from_conn, from_addr) = listen_socket.accept().await
+                .chain_err(|| 
+                    format!("error accepting connection")
+                )?;
+            println!("connection received from {}", from_addr);
+
+            let to_addr = server_carousel.next()
+                .chain_err(|| "server carousel failed to provide server addr")?
+                .clone();
+
+            tokio::spawn(async move {
+                if let Ok(to_conn) = TcpStream::connect(to_addr).await {
+                    println!("connection opened to {}", to_addr);
+                    server(from_conn, to_conn, compress, encrypt).await;
+                }
+                else {
+                    eprintln!("failed to connect to {}", to_addr);
+                }
+            });
+        }
+    })
+}
+
+async fn server(
+    mut from_conn: TcpStream,
+    mut to_conn: TcpStream,
+    _compress: bool,
+    _encrypt: bool
+)
+{
+    let mut buf = vec![0; 1024];
+
+    loop {
+        // echo client to server
+        match from_conn.read(&mut buf).await {
+            Ok(0) => {
+                println!("Client closed connection");
+                break;
+            }
+            Ok(n) => {
+                if let Err(_) = to_conn.write_all(&buf[..n]).await {
```

Comment:
What happens if n > buffer size (like a get request with a reply of 10KB)? I'm not familiar enough to know if it will read and send multiple times.

Reply:
Rust does runtime bounds checking so an exception would be raised if n was greater than the buffer size here (as long as you don't wrap your code in an unsafe block). 

write_all() will continuously write chunks until the whole buffer has been written, so it takes care of multiple requests for you. It is definitely possible that it will send the buffer out over more (or fewer) packets than it came in on, but that's all handled under the hood.

Similarly, read is supposed to guarantee that it will never overflow the buffer. From the read interface documentation: 

> If the return value of this method is Ok(n), then it must be guaranteed that 0 <= n <= buf.len().

So if you had a larger message that you wanted to read all of, you would have to call read multiple times.

Reply:
Ah, ok. And I'm guessing Ok(n) means that n is the "Content-Length" header, right? Or is that how much is received by the socket? I think it would make a difference.

I think I thought about this in my PA2. What I did was somehing like
bytesRemaining = n;
while(bytesRemaining > 0) {
    btoread = min(bufferSpace, bytesRemaining);
    read(btoread);
    bytesRemaining -= btoread;
}

I don't remember who wrote the code but I think this is something we may want to do, right?
And also bufferSpace would not be the capacity of the buffer, but however much room it has (in case the write doesn't write it all out at once), so you may have to check how many bytes you successfully write as well. Though it does sound like the write_all() might guarantee that the whole buffer is written out before the next read.

Reply:
n is the number of bytes read into buffer by the read() call. Since we're reading from a TCP stream, it doesn't have any understanding of the HTTP data it's sending, so it doesn't understand the content length header. An http response that looks something like: 
```
HTTP/1.1 200 OK
Content-Length: 11

Hello World
```

Might be read in two separate chunks as:
```
HTTP/1.1 200 OK
Content-Leng
```
and
```
th: 11

Hello World
```

TCP just treats it as a stream and will try to continuously write and/or read whatever data comes in over the stream. In this case, the read calls are in an infinite loop so we don't need to worry about checking for remaining data since if there's more data to read and send, it will be picked up by the next read call in the infinite loop